pkg/v1/sdk/capabilities/discovery/cluster_gvr.go

```diff
@@ -35,94 +58,239 @@ func (q *QueryGVR) Name() string {
+	groupList, err := cfg.discoveryClientset.ServerGroups()
 	if err != nil {
-		return false, err
+		return nil, fmt.Errorf("discovering server groups: %w", err)
 	}
 
 	var unmatched []string
-	for _, gvr := range gvrs {
-		// Check group.
-		if !q.groupExists(gvr.Group, groupList) {
-			unmatched = append(unmatched, gvr.String())
-			continue
+
+	group := q.groupFromGroupList(groupList)
+	if group == nil {
+		// No group version matches because group could not be found.
+		for _, version := range q.versions {
+			unmatched = append(unmatched, schema.GroupVersionResource{
+				Group:    q.group,
+				Version:  version,
+				Resource: q.resource.String,
+			}.String())
 		}
+		return unmatched, nil
+	}
 
-		// Check version if specified.
-		if gvr.Version == "" {
-			continue
+	// Group was found. Find which query versions are not.
+	for _, queryVersion := range q.versions {
+		var matched bool
+		for _, discovery := range group.Versions {
+			if strings.EqualFold(discovery.Version, queryVersion) {
+				matched = true
+				continue
+			}
 		}
+		if !matched {
+			unmatched = append(unmatched, schema.GroupVersionResource{
+				Group:    q.group,
+				Version:  queryVersion,
+				Resource: q.resource.String,
+			}.String())
+		}
+	}
+	return unmatched, nil
+}
 
-		resources, err := config.discoveryClientset.ServerResourcesForGroupVersion(gvr.GroupVersion().String())
+func (q *QueryGVR) unmatchedGroupVersionResources(cfg *clusterQueryClientConfig) ([]string, error) {
+	var unmatched []string
+	for _, ver := range q.versions {
+		gvr := schema.GroupVersionResource{
+			Group:    q.group,
+			Version:  ver,
+			Resource: q.resource.String,
+		}
+		resources, err := cfg.discoveryClientset.ServerResourcesForGroupVersion(gvr.GroupVersion().String())
 		if err != nil {
 			// Second condition is because fake discovery client does not return a proper NotFound error.
-			if apierrors.IsNotFound(err) || strings.Contains(err.Error(), fmt.Sprintf("GroupVersion %q not found", gvr.GroupVersion().String())) {
+			if apierrors.IsNotFound(err) || strings.Contains(
+				err.Error(),
+				fmt.Sprintf("GroupVersion %q not found", gvr.GroupVersion().String()),
+			) {
 				unmatched = append(unmatched, gvr.String())
-				continue
+			} else {
+				return nil, err
 			}
-			return false, err
-		}
-
-		// Check resource if specified.
-		if gvr.Resource == "" {
-			continue
 		}
 		if !q.resourceExists(resources) {
 			unmatched = append(unmatched, gvr.String())
-			continue
 		}
 	}
-
-	q.unmatchedGVRs = unmatched
-	return len(unmatched) == 0, nil
+	return unmatched, nil
 }
 
-func (q *QueryGVR) groupExists(group string, groupList *metav1.APIGroupList) bool {
-	for i := range groupList.Groups {
-		if strings.EqualFold(groupList.Groups[i].Name, group) {
+// groupsContains looks for a group in a slice of APIGroup.
+func (q *QueryGVR) groupsContains(groups []*metav1.APIGroup, group string) bool {
```

Comment:
nit: `s/groupsContains/containsGroup`

Reply:
I chose `groupsContains` because it reads similar to the standard library method `strings.Contains`. I even ordered the arguments to be first the haystack and the needle (groups first, then specific group second).

Since it's a nit, I'll leave it how it is. However, if someone has a strong opinion, I don't mind changing it to `containsGroup`.

Reply:
I vote for `containsGroup` as well

Reply:
Changed `groupsContains` -> `containsGroup`.